src/net_processing.cpp

```diff
@@ -3358,7 +3375,13 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             // requests for it.
             m_txrequest.ForgetTxHash(tx.GetHash());
             m_txrequest.ForgetTxHash(tx.GetWitnessHash());
-            RelayTransaction(tx.GetHash(), tx.GetWitnessHash(), m_connman);
+
+            // Flood those transactions which were received either via flooding, or inbound reconciliation,
+            // but NOT via outbound reconciliation. Flooding then is mainly used for initial propagation
+            // of new transactions across a network of reachable nodes quickly.
+            LOCK(peer->m_recon_state_mutex);
+            bool flood = !(peer->m_recon_state && peer->m_recon_state->m_responder);
```

Comment:
Does this work as intended in the intermediate stage where not all of our peers support Erlay? If we are a non-reachable node that supports Erlay but has just one older pre-Erlay outbound peer - wouldn't we flood all of its transactions also to our Erlay-supporting peers because `flood == true` here? (Considering the goal that "only well-connected publicly reachable nodes flood transactions to other publicly reachable nodes via outbound connections").

Reply:
"only well-connected publicly reachable nodes flood transactions to other publicly reachable nodes via outbound connections" only is an end-goal when most of the network implement Erlay.

In the meanwhile, I think this behavior is better because it's most intuitive: for legacy peers, we support legacy behavior (flood).
Technically, changing this condition wouldn't break stuff, but I think it's just less intuitive.

I probably should expand the comment to address this misunderstanding though.

Reply:
But the legacy behavior also affects Erlay peers (because if a non-reachable node received transactions from legacy  peers, it would act as if it was a reachable hub and forward them via flooding to other outbound peers, including Erlay ones).

Apart from this specific spot that could easily be adjusted if needed, I think that when there are two relay mechanisms with different time scales existing in parallel, this could lead to some non-intuitive consequences - such as legacy relay being dominant and cannibalizing Erlay relay way into the regime when the majority of nodes already support Erlay.
 
Would it maybe be worth doing a mixed-network simulation run and measuring the respective contributions at different ratios of Erlay to legacy nodes, if this is possible within the existing simulation framework?

Reply:
@mzumsande sorry I just saw this question. Yeah, that might be useful, once I'm done with big refactor for this...
Intuitively, I think this shouldn't be the case legacy peers cannibalize efficiency I think.